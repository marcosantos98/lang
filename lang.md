# Lang

Number :: 19293823
Str :: "dasdas"
Char :: 'c'
Bool :: true | false
Ident :: [a-zA-Z_]

Literal :: Number | Str | Char | Bool

Stmt :: Literal | FnCall

FnCall :: Ident "(" [Stmt] ")"

BlockExpr :: "{" Stmt "}"
FnDeclExpr :: Ident "::" Ident "(" [Ident,Ident]* ")" BlockExpr
ExternFnDeclExpr :: "extern" Str Ident "::" Ident "(" [Ident Ident]* ")"
VarDeclExpr :: Ident ":=" Stmt
ImportExpr :: "import" Str

## fn

```
sum :: int(a int, b int) {
    return a + b
}

main :: void() {
}

extern "c" printf :: int(str cstr);
```

## literals

```
string -> ""
char -> ''
number -> 19382
bool -> true | false
```

## import

Import tokenizes and parses the given file and outputs everthing to the main translation.
Currently it copies everthing to who imported it.

```
import "<filepath>"
```

## var

```
a := 0
b := ""
c := fn()
```
