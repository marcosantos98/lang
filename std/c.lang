extern "c" puts :: int(str char*)
extern "c" putchar :: int(d int)
extern "c" exit :: int(code int)
extern "c" printf :: int(str char*, args ...)
