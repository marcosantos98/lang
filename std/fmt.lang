import "std:c"

putd :: void(d int) {
    if d > 9 {
	var a = d / 10
	d -= 10 * a
	putd(a)
    }
    putchar(48 + d)
}
