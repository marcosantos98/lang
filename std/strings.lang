import "std:c"

struct StringBuilder {
    data ^char,
    len int,
    cap int,
    a string
}

sb_new :: StringBuilder() {
    var data = malloc(1024)
    memset(data, 0, 1024)
    return StringBuilder{data as ^char, 0, 1024}
}

sb_write_str :: void(sb ^StringBuilder, append string) {
    var new_len = sb.len + len(append)
    if new_len > sb.cap {
	sb.cap = sb.cap * 2
	sb.data = realloc(sb.data, sb.cap) as ^char
    }
    var ptr: ^char = sb.data as ^char
    memcpy(ptr + sb.len, append.data, len(append))
    sb.len = new_len
}

sb_print :: void(sb StringBuilder) {
    printf(c("%.*s"), sb.len, sb.data)
}

sb_to_cstr :: ^char(sb StringBuilder) {
    return sb.data as ^char
}

sb_destroy :: void(sb StringBuilder) {
    free(sb.data)
}
