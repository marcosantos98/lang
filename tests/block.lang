import "std.lang"

main :: void(){
    a := "Hello, World"
    {
	puts(a)
    }
}
