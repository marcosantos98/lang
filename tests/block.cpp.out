/* -- Builtin types -- */
typedef const char* cstr;
/* -- ------------- -- */

extern "C" int puts(cstr);
extern "C" int exit(int);
void ___entry___();

void ___entry___(){
cstr a = "Hello, World";
{
puts(a);
}
}
int main(int argc, char** argv) {___entry___();}
